### PR TITLE
TestSignalGenerator Module

### LINKED ISSUE #221: Implement a Test Signal Generator Module
**Is your feature request related to a problem? Please describe.**
Implement a test signal generator source module which generates a test pattern for the pipeline.  

**Describe the solution you'd like**
The source process will write a test pattern into a buffer and send in frame.
Eg: A for loop writes a YUV420 test pattern  into a frame. 

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Vinayak-YB**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #221 

**Description**

This module generates test signal for YUV420 image,which can be used for testing purpose.

**Type**

Type Choose one: (Bug fix | Feature | Documentation | Testing | Other)

**Screenshots (if applicable)**

**Checklist**

- [x] I have read the [Contribution Guidelines](https://github.com/Apra-Labs/ApraPipes/wiki/Contribution-Guidelines)
- [x] I have written Unit Tests
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach

### CODE REVIEW

File: base/test/test_utils.cpp [hunk 1]
```diff
@@ -110,12 +112,34 @@ bool Test_Utils::saveOrCompare(const char* fileName, const unsigned char* dataTo
 
 	if (boost::filesystem::is_regular_file(fileName))
 	{
+		const std::string strFullFileName = fileName;
+		std::string strFileBaseName = strFullFileName;
+		std::string strExtension;
+		std::string nameOfFile;
+
+		const size_t fileBaseIndex = strFullFileName.find_last_of(".");
+		const size_t fileBaseLocation = strFullFileName.find_last_of("/");
```
Comment (mraduldubey) [inline base/test/test_utils.cpp:121]: test with simple filename with no / eg. file.raw

File: base/test/test_utils.cpp [hunk 2]
```diff
@@ -110,12 +112,34 @@ bool Test_Utils::saveOrCompare(const char* fileName, const unsigned char* dataTo
...
+		if (std::string::npos != fileBaseIndex)
+		{
+			strFileBaseName = strFullFileName.substr(0U, fileBaseIndex);
+			strExtension = strFullFileName.substr(fileBaseIndex);
+		}
+		if (std::string::npos != fileBaseLocation)
```
Comment (mraduldubey) [inline base/test/test_utils.cpp:127]: add small comments here

File: base/test/test_utils.cpp [hunk 3]
```diff
@@ -110,12 +112,34 @@ bool Test_Utils::saveOrCompare(const char* fileName, const unsigned char* dataTo
...
+
 		const uint8_t* dataRead = nullptr;
 		unsigned int dataSize = 0U;
 		BOOST_TEST(readFile(fileName, dataRead, dataSize));
 		BOOST_TEST(dataSize == sizeToSC);
-
 		compareRes = CompareData(dataToSC, dataRead, dataSize, tolerance);
+		if(!compareRes)
+		{
+			boost::filesystem::create_directory("./data/SaveOrCompareFail");
```
Comment (mraduldubey) [inline base/test/test_utils.cpp:139]: will create_directory throw an error or silently fail if the directory already exists ?

**Review by mraduldubey — CHANGES_REQUESTED** [review]

File: base/test/test_utils.cpp (hunk 1 above)
Comment (Vinayak-YB) [reply base/test/test_utils.cpp:122]: Tested with only TestSample.raw it worked

File: base/test/test_utils.cpp (hunk 2 above)
Comment (Vinayak-YB) [reply base/test/test_utils.cpp:128]: Added a few comments

File: base/test/test_utils.cpp (hunk 3 above)
Comment (Vinayak-YB) [reply base/test/test_utils.cpp:140]: If directory already exists it will fail silently

**Review by mraduldubey — APPROVED** [review]

**Pull request merged**